submissions/api.py

```diff
@@ -637,12 +639,69 @@ def get_latest_score_for_submission(submission_uuid, read_replica=False):
         score = score_qs[0]
         if score.is_hidden():
             return None
-    except IndexError:
+    except (IndexError, Submission.DoesNotExist):
         return None
 
     return ScoreSerializer(score).data
 
 
+def reset_student_item(student_id, course_id, item_id):
+    """
+    In order to reset the student's progress on an item, we soft-delete any associated submissions. Since all
+    interaction with scores comes via the submission's uuid, this has the effect of making the student's state
+    on this item be reset to a fresh state.
+    """
+    try:
+        # Retrieve the student item
+        student_item = StudentItem.objects.get(
+            student_id=student_id, course_id=course_id, item_id=item_id
+        )
+
+        # There will *usually* only be one submission, but that's not guaranteed so "delete" all items
+        submissions = student_item.submission_set.all()
+        ret_vals = [sub.uuid for sub in submissions]
+        for sub in submissions:
+            cache_key = "submissions.submission.{}".format(sub.uuid)
+            cache.delete(cache_key)
+            sub.deleted = True
+            sub.save()
+
+        # ScoreSummary and StudentItem have uniqueness constraints that make soft-deleting problematic
+        # Luckily, creating a reset score will handle our needs adequately
+        Score.create_reset_score(student_item)
+
+        # TODO: the top scores cache is automatically invalidated after 5 minutes, should we rely on that instead of doing it manually here?
```

Comment:
Open question for reviewers here, the manual cache clearing is a bit unwieldy since `number` is included in the keys, is it worth doing anyways or should we let the cache expire?